Arm64 Image Release Blog (#243)
aws-otel/aws-otel.github.io: src/content/Blogs/arm64Images.mdx

---
title: 'Adding ARM64 support to AWS Distro for OpenTelemetry (ADOT) Docker containers'
description:
    This Blog post is the release announcement for arm64 docker images
---

**By Seth Levine, ADOT Team**

import SectionSeparator from "components/MdxSectionSeparator/sectionSeparator.jsx"

import seth from "assets/img/blogs/seth.jpg"

<SectionSeparator />

[AWS Distro for OpenTelemetry (ADOT)](https://aws-otel.github.io/) arm64 images is now available. You can download the latest [ADOT Collector image](https://gallery.ecr.aws/aws-observability/aws-otel-collector) from the [Amazon Elastic Container Registry (Amazon ECR)](https://aws.amazon.com/ecr/) Public Gallery.

<SectionSeparator />

With the release of the AWS Distro for OpenTelemetry (ADOT) Collector agent 0.16.0, customers will be able to also run the Collector agent  on arm64 docker containers.

> You can Locate ARM64 images [here](https://gallery.ecr.aws/aws-observability/aws-otel-collector). Since docker registries allow multiple manifests per tag based on the architecture, you will not need to change your docker-compose / Kube deployment files when transitioning to arm64 ec2 instances (more information later in the blog post).

Reviewer: ```suggestion
You can Locate ARM64 images [here](https://gallery.ecr.aws/aws-observability/aws-otel-collector). Since container registries allow multiple manifests per tag based on the architecture, you will not need to change your docker-compose / Kubernetes deployment files when transitioning to ARM64 hosts (more information later in the blog post).
```

**What was already supported, and what needed to change?**

> We already supported the amd64 and arm64 binary for Linux, but we only used docker build to create containers.

Reviewer: ```suggestion
We already supported the AMD64 and ARM64 binary for Linux, but we used `docker build` to create containers.
```

> Docker build will only create containers for the default architecture of the system. For our that ci that is amd64. We needed a solution that supported arm64 images. We decided to go with docker buildx, which allows multi-architecture builds. See docs [here](https://docs.docker.com/buildx/working-with-buildx/).

Reviewer: ```suggestion
Docker's `build` command will only create containers for the architecture of the host system. For our that build systems that is aMD64. We needed a solution that supported ARM64 images. We decided to go with `docker buildx`, which allows multi-architecture builds. See docs [here](https://docs.docker.com/buildx/working-with-buildx/).
```

**We are changing the docker file.**

> When running the Dockerfile via the docker build command, we only had access to the amd64 binary. We added the ability to access the arm64 binary inside the docker build container. The docker build command gives an argument platform that translates to TARGETARCH in the Dockerfile. Using the knowledge of the docker target platform arch, we could choose which binary to copy to the docker build container. See Dockerfile [here](https://github.com/aws-observability/aws-otel-collector/blob/main/cmd/awscollector/Dockerfile)

Reviewer: ```suggestion
When running the `Dockerfile` via the `docker build` command, we only had access to the AMD64 binary. We added the ability to access the ARM64 binary inside the docker build container. The `docker build` command gives an argument platform that translates to `TARGETARCH` in the `Dockerfile`. Using the knowledge of the docker target platform architecture, we could choose which binary to copy to the docker build container. See Dockerfile [here](https://github.com/aws-observability/aws-otel-collector/blob/main/cmd/awscollector/Dockerfile)
```

This should be further reworded, it's still somewhat confusing.  Do we really need to go into this level of detail?  Maybe we can just say "we did the thing!" and people can look at the commit history if they care how we did the thing.

> **Changing the make file**
> 
> Since local docker only supports loading single image manifest, we only build for single architectures when not pushing to remote. In the make file, we split the amd64 image and arm64 image into two separate commands seen [here](https://github.com/aws-observability/aws-otel-collector/blob/main/Makefile#L112-L118).
> 
> **Changing GitHub actions**
> 
> In our ci, we currently only built and uploaded amd64 images. With the GitHub actions for docker, we could specify multiple platform build and push docks [here](https://github.com/marketplace/actions/build-and-push-docker-images). For our release process, we used to load the image locally then push it up to the public ECR. We could not do this because images were now multiple manifests. So we went with GitHub actions copy docker actions docs [here](https://github.com/docker/build-push-action/blob/master/docs/advanced/copy-between-registries.md).
> 
> **Testing**
> 
> We wanted to make sure the arm64 image worked the same as the amd64 image. We created an arm64 EKS cluster. Then we updated all our test apps to support arm64 docker images. Then we ran our EKS tests against the arm64 cluster. We ensured that the tests returned the same values in both architectures. You can find our [integration tests](https://github.com/aws-observability/aws-otel-collector/blob/main/e2etest/testcases.json), [testing framework](https://github.com/aws-observability/aws-otel-test-framework), and [results](https://github.com/aws-observability/aws-otel-collector/actions/workflows/CI.yml).
> 

Reviewer: ```suggestion
```

I don't think we need to go into this level of detail.

Reviewer 2: I guess the initial idea for this blog post was to announce and include the implementation details. Yes, this level of detail might not feel relevant for a blog post. but for info under this **What was already supported, and what needed to change?** says this post is trying to include the implementation part too.
